### PR TITLE
Domain Picker: refactor folder structure

### LINKED ISSUE #51144: Domain Picker: rework folder structure
## What

Rearrange the folder structure of the `src` folder in the Domain Picker package

## Why

This will make the package more structured and easier to work with, in addition to enabling us to leave each unit test next to its source file.

## A/C

Folder structure spec:

```
src/
    - __mocks__/
    |   - common.mocks.data.ts
    - __tests__/
    |   - integration.test.tsx
    - components/
    |   - __tests__/
    |   |   - unit.test.tsx
    |   - index.tsx (actual domain picker root component)
    |   - style.scss
    |   - subcomponent-a/
    |   |   - __mocks__/
    |   |   |   - specific.mock.data.needed.only.here.ts
    |   |   - __tests__/
    |   |   |   - unit.test.tsx
    |   |   - index.tsx
    |   |   - style.scss
    |   - subcomponent-b/
    |   |   - __tests__/
    |   |   |   - unit.test.tsx
    |   |   - index.tsx
    |   |   - style.scss
    - hooks/
    |   - __tests__/
    |   |   - unit.test.tsx
    |   - index.ts
    |   - [other hook files]
    - utils/
    |   - __tests__/
    |   |   - unit.test.ts
    |   - index.ts
    - [other files]
```

- [x] The Domain Picker folder structure should follow the spec highlighted above.
- [x] All React components should be moved under the `components` folder.
- [x] Each component (or sub-component) folder should have an `index.tsx` file (and a `style.scss` when applicable).
- [x] The domain suggestion item components should be moved to a separate sub-component folder
- [x] For the purpose of this issue, we should disable the linter rule forcing CSS classnames based on a file path
- [x] Each component/subcomponent folder should have a `__test__` folder, where unit tests for the components in that folder are stored (we can start with a dummy test file for this issue)
- [x] Integration tests are localed in the `src/__tests__` folder
- [x] Common mock data is stored in the `src/__mocks__` folder
- [x] Mock data that is specific to a single unit test should be store in a `__mocks__` folder that is a sibling of the `__tests__` folder containing the unit tests

## Context

Restructuring the folder structure will allow for a clear distinction between unit and integration tests.

The styles in `packages/domain-picker/src/domain-picker/style.scss` should be split into more files, separating the styles for the domain suggestion items from the rest of the styles — tracked in separately in #51187. We should also re-enable the linter rule about classnames following the file path.

### PR DESCRIPTION
**Pull request by ciampo**
#### Changes proposed in this Pull Request

* Refactor the structure of the Domain Picker `src` folder to allow for better modularity and tests granularity
* Refer to #51144 and to this PR's commit log for details list of changes

#### Testing instructions

These changes should only affect the internal structure of the Domain Picker package, and therefore shouldn't affect any user-facing behaviour

* Run `yarn && yarn start` and make sure that the project builds correctly
* Run `cd apps/editing-toolkit && yarn build` and make sure that the project builds correctly
* Run `cd packages/domain-picker && yarn test` and make sure that all tests pass (the `<button> cannot appear as a descendant of <button>.` error is expected)
* Smoke test Gutenboarding and Focused Launch

Closes #51144

### CODE REVIEW

File: packages/domain-picker/jest.config.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 module.exports = {
 	preset: '../../test/packages/jest-preset.js',
 	testEnvironment: 'jsdom',
+	testMatch: [ '<rootDir>/**/__test__/**/*.[jt]s?(x)', '!**/.eslintrc.*' ],
```
Comment (ciampo) [inline packages/domain-picker/jest.config.js]: Override default Calypso's config, so that we can put our tests in folders named `__test__`

File: packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx [hunk 2]
```diff
@@ -12,8 +8,8 @@ import { render, fireEvent, screen } from '@testing-library/react';
  * Internal dependencies
  */
 // https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
-import '../matchMedia.mock';
-import SuggestionItem from '../src/domain-picker/suggestion-item';
+import '../../../__mock__/matchMedia.mock';
+import SuggestionItem from '../suggestion-item';
```
Comment (ciampo) [inline packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx]: @StefanNieuwenhuis FYI — this will likely cause merge conflicts with #51177

File: packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx [hunk 3]
```diff
@@ -1,7 +1,3 @@
-/**
- * @jest-environment jsdom
- */
-
```
Comment (ciampo) [inline packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx]: Not necessary since it's already specified in this package's jest config

File: packages/domain-picker/src/hooks/__test__/use-domain-suggestions.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+describe( 'useDomainSuggestions hook', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Comment (ciampo) [inline packages/domain-picker/src/hooks/__test__/use-domain-suggestions.ts]: Opened #51242 to track Domain Picker hooks unit tests

File: packages/domain-picker/src/hooks/__test__/use-persistent-selected-domain.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+describe( 'usePersistentSelectedDomain hook', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Comment (ciampo) [inline packages/domain-picker/src/hooks/__test__/use-persistent-selected-domain.ts]: Opened #51242 to track Domain Picker hooks unit tests

File: packages/domain-picker/src/utils/__test__/is-good-default-domain-query.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 /**
  * Internal dependencies
  */
-import { isGoodDefaultDomainQuery } from '../src/utils';
+import { isGoodDefaultDomainQuery } from '../';
```
Comment (ciampo) [inline packages/domain-picker/src/utils/__test__/is-good-default-domain-query.ts]: Missing utils unit tests tracked in #51243

File: packages/domain-picker/src/components/__test__/domain-picker.tsx [hunk 7]
```diff
@@ -0,0 +1,5 @@
+describe( '<DomainPicker />', () => {
+	it( 'should be replaced with real tests', () => {
+		expect( 1 ).toBeTruthy();
+	} );
+} );
```
Comment (ciampo) [inline packages/domain-picker/src/components/__test__/domain-picker.tsx]: Missing domain picker components unit tests tracked in #51244

File: packages/domain-picker/jest.config.js (hunk 1 above)
Comment (StefanNieuwenhuis) [reply packages/domain-picker/jest.config.js]: Maybe a bit of [nitpicking/] since Calypso is overruling the default Jest config anyway, but Jest's default is `__tests__` (plural) (see: https://jestjs.io/docs/configuration#testmatch-arraystring). The same goes for `__mock__` vs. `__mocks__`.

WDYT? Do we need to use plural too?

File: packages/domain-picker/src/__mock__/suggestions.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+/**
+ * External dependencies
+ */
+import type { DomainSuggestion } from '@automattic/data-stores/src/domain-suggestions';
+
+export const MOCK_DOMAIN_SUGGESTION: DomainSuggestion = {
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/__mock__/suggestions.ts]: I think we have to extend and use this mock in #51177 too.

File: packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx (hunk 2 above)
Comment (StefanNieuwenhuis) [reply packages/domain-picker/src/components/domain-suggestion-item/__test__/suggestion-item.tsx]: It certainly will and we have to use the DomainSuggestions mock there as well. So we have to discuss our merging strategy. 

I opt for merging this one first and rebase, update & merge #51177 after.

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx [hunk 9]
```diff
@@ -6,7 +6,7 @@ import * as React from 'react';
 /**
  * Internal dependencies
  */
-import type { SUGGESTION_ITEM_TYPE } from 'src';
+import { SUGGESTION_ITEM_TYPE, SUGGESTION_ITEM_TYPE_BUTTON } from './suggestion-item';
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx]: What do you think about this, more explicit, though verbose, approach?

```suggestion
import type { SUGGESTION_ITEM_TYPE } from './suggestion-item';
import { SUGGESTION_ITEM_TYPE_BUTTON } from './suggestion-item';
```

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx [hunk 10]
```diff
@@ -20,7 +20,7 @@ type WrappingComponentProps = WrappingComponentAdditionalProps &
 
 const WrappingComponent = React.forwardRef< HTMLButtonElement, WrappingComponentProps >(
 	( { type, disabled, ...props }, ref ) => {
-		if ( type === 'button' ) {
+		if ( type === SUGGESTION_ITEM_TYPE_BUTTON ) {
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx:25]: [nitpick/] This is scope creep since it doesn't have anything to do with folder restructuring.

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx [hunk 11]
```diff
@@ -23,15 +25,16 @@ import { sprintf } from '@wordpress/i18n';
...
+export const SUGGESTION_ITEM_TYPE_BUTTON = 'button';
+export const SUGGESTION_ITEM_TYPE_INDIVIDUAL = 'individual-item';
 export type SUGGESTION_ITEM_TYPE =
-	| typeof ITEM_TYPE_RADIO
-	| typeof ITEM_TYPE_BUTTON
-	| typeof ITEM_TYPE_INDIVIDUAL_ITEM;
+	| typeof SUGGESTION_ITEM_TYPE_RADIO
+	| typeof SUGGESTION_ITEM_TYPE_BUTTON
+	| typeof SUGGESTION_ITEM_TYPE_INDIVIDUAL;
+
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:37]: [nitpick/] This is scope creep since it doesn't have anything to do with folder restructuring.

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx [hunk 12]
```diff
@@ -23,15 +25,16 @@ import { sprintf } from '@wordpress/i18n';
 import InfoTooltip from '../info-tooltip';
 import WrappingComponent from './suggestion-item-wrapper';
 // TODO: remove when all needed core types are available
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:27]: Do we need an issue for this?

**Review by StefanNieuwenhuis — COMMENTED** [review]
Thanks for this PR @ciampo! It looks so structured already and I've tested it successfully! 

My only concern is the placeholder unit tests. Since we're not sure if we need to create unit tests for all (sub)components or that we can cover them with integration tests, I think that they should be left out. WDYT?

I left some inline comments as well. Can you please address them?

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx (hunk 10 above)
Comment (ciampo) [reply packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx:25]: That's partially true, although the reason these changes are in this PR is because initially I had extracted these variables to a separate file during the folder restructure, and then I changed my mind and moved them back.

It's a small scope creek but probably we can keep it in, since the changes are not affecting the functionality of the component?

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx (hunk 11 above)
Comment (ciampo) [reply packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:37]: See https://github.com/Automattic/wp-calypso/pull/51235#discussion_r599720224

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx (hunk 12 above)
Comment (ciampo) [reply packages/domain-picker/src/components/domain-suggestion-item/suggestion-item.tsx:27]: Opened #51328 to track it

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx (hunk 9 above)
Comment (ciampo) [reply packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx]: Done in https://github.com/Automattic/wp-calypso/pull/51235/commits/693ce38a124aa02293aeb15367692937c823995f

File: packages/domain-picker/jest.config.js (hunk 1 above)
Comment (ciampo) [reply packages/domain-picker/jest.config.js]: Updated the original issue's A/C and done across these commits:
- https://github.com/Automattic/wp-calypso/pull/51235/commits/cd951f36e0d91f5c4daea4f8571c9c81caf950f5
- https://github.com/Automattic/wp-calypso/pull/51235/commits/a5c22b64f101483b8e32811a250a4865bed11d47
- https://github.com/Automattic/wp-calypso/pull/51235/commits/eea4c35d2ebe964020cbecdd29e5759b8566563a
- https://github.com/Automattic/wp-calypso/pull/51235/commits/512ee17ce81e4ea1c82ab32dd4b95fd2d59e108d

**Comment by ciampo** [comment]
Hey @StefanNieuwenhuis, I addressed all of your comments. Can you have another look?


> My only concern is the placeholder unit tests. Since we're not sure if we need to create unit tests for all (sub)components or that we can cover them with integration tests, I think that they should be left out. WDYT?

Removed placeholder tests in https://github.com/Automattic/wp-calypso/pull/51235/commits/99a2df3e926af3b361c092537f53987df20f7812

File: packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx (hunk 10 above)
Comment (StefanNieuwenhuis) [reply packages/domain-picker/src/components/domain-suggestion-item/suggestion-item-wrapper.tsx:25]: Thanks for elaborating. Although we have to consider it as scope creep, we can keep it in since the changes are not affecting the functionality of the component

**Review by StefanNieuwenhuis — APPROVED** [review]
Code & Tests LGTM! 🚢
